lerna-util-sequence/src/main/scala/lerna/util/sequence/SequenceFactoryWorker.scala

```diff
@@ -128,127 +128,136 @@ private[sequence] final class SequenceFactoryWorker(
+    case Initialize                                                             => Behaviors.unhandled
+    case WrappedSequenceStoreResponse(_: SequenceStore.InitialSequenceReserved) => Behaviors.unhandled
+    case WrappedSequenceStoreResponse(msg: SequenceStore.SequenceReset)         =>
+      // reset するときは必ず empty になっているため
+      Behaviors.unhandled
+  }
 
-  private[this] def resetting: Behavior[Command] = Behaviors.receiveMessage[Command] {
+  private[this] def empty(sequenceContext: SequenceContext): Behavior[Command] = Behaviors.receiveMessage {
+    case msg: GenerateSequence =>
+      // no reply
+      logger.warn(
+        "Pending generate sequence because reserving sequence: current max reserved: {}, next sequence value: {}",
+        sequenceContext.maxReservedValue,
+        sequenceContext.nextValue,
+      )
+      stashBuffer.stash(msg)
+      prepareNextSequence(sequenceContext = sequenceContext, nextSequence = sequenceContext)
+    case WrappedSequenceStoreResponse(msg: SequenceStore.SequenceReserved) =>
+      stashBuffer.unstashAll(handleSequenceReserved(msg, sequenceContext))
     case WrappedSequenceStoreResponse(msg: SequenceStore.SequenceReset) =>
-      logger.warn("reset sequence: {}", msg.maxReservedValue)
-      stashBuffer.unstashAll(ready(SequenceContext(msg.maxReservedValue, nextValue = firstValue)))
-    case message: GenerateSequence =>
-      stashBuffer.stash(message)
-      Behaviors.same
+      stashBuffer.unstashAll(handleSequenceReset(msg, sequenceContext))
+    case WrappedSequenceStoreResponse(SequenceStore.ReservationFailed)          => Behaviors.same
     case ReceiveTimeout                                                         => Behaviors.stopped
     case Initialize                                                             => Behaviors.unhandled
     case WrappedSequenceStoreResponse(_: SequenceStore.InitialSequenceReserved) => Behaviors.unhandled
-    case WrappedSequenceStoreResponse(_: SequenceStore.SequenceReserved)        => Behaviors.unhandled
-    case WrappedSequenceStoreResponse(SequenceStore.ReservationFailed)          => Behaviors.unhandled // FIXME
+  }
+
+  private[this] def acceptGenerateSequence(
+      msg: GenerateSequence,
+      sequenceContext: SequenceContext,
+  ): Behavior[Command] = {
+    msg.replyTo ! SequenceGenerated(sequenceContext.nextValue, sequenceSubId)
+    logger.debug("SequenceGenerated: {}", sequenceContext.nextValue)
+    prepareNextSequence(sequenceContext = sequenceContext, nextSequence = sequenceContext.next())
+  }
+
+  private[this] def prepareNextSequence(
+      sequenceContext: SequenceContext,
+      nextSequence: SequenceContext,
+  ): Behavior[Command] = {
+    if (nextSequence.isOverflow) {
+      reset()
+      empty(nextSequence)
+    } else if (nextSequence.isEmpty) {
+      empty(nextSequence)
+    } else if (nextSequence.isStarving) {
+      reserve(sequenceContext = sequenceContext, nextSequence = nextSequence)
+      ready(nextSequence)
```

Comment:
`def prepareNextSequence`
```scala
    } else if (nextSequence.isEmpty) {
      empty(nextSequence)
    } else if (nextSequence.isStarving) {
      reserve(sequenceContext = sequenceContext, nextSequence = nextSequence)
      ready(nextSequence)
```

isEmpty の場合、 isStarving を満たしますが、reserve はしないですか？

変更前は `def reserving` では同時に1つしか予約しないようになっていました。
しかし、今は複数同時に予約するようになっているので、ここだけ予約しないのは変に感じます。

> SequenceFactoryWorker が SequenceStore の状態によらず採番値を予約するようになった

```scala
          } else if (sequenceContext.isEmpty) {
            logger.warn("Pending generate sequence because reserving sequence: {}", sequenceContext.nextValue)
            stashBuffer.stash(msg)
            Behaviors.same
```

Reply:
特定条件で採番機が停止しますね。

```scala
    "採番値が枯渇しているときに予約に失敗した場合は次の採番要求で再度予約が要求される" in {
      val storeProbe        = createTestProbe[SequenceStore.Command]()
      val firstValue        = 3
      val incrementStep     = 10
      val reservationAmount = 1 // 2以上なら問題ない
      val worker = spawn(
        SequenceFactoryWorker
          .apply(
            maxSequenceValue = 999,
            firstValue = firstValue,
            incrementStep = incrementStep,
            reservationAmount = reservationAmount,
            storeProbe.ref,
            idleTimeout = 10.seconds,
            sequenceSubId,
          ),
      )
      // 初期化
      inside(storeProbe.receiveMessage()) {
        case result: SequenceStore.InitialReserveSequence =>
          result.replyTo ! SequenceStore.InitialSequenceReserved(
            initialValue = firstValue,                                                // 3
            maxReservedValue = firstValue + (incrementStep * (reservationAmount - 1)),// 3
          )
      }

      // 採番要求: reservationAmount が 1 なので、1回採番すると枯渇 → 予約
      {
        val replyToProbe = createTestProbe[SequenceFactoryWorker.SequenceGenerated]()
        worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref)
        expect(replyToProbe.receiveMessage().value === BigInt(3)) // firstValue
      }

      // 予約を失敗させる
      inside(storeProbe.receiveMessage()) {
        case result: SequenceStore.ReserveSequence =>
          result.replyTo ! SequenceStore.ReservationFailed
      }

      // 採番要求: 予約できていないのでレスポンスは来ない
      val replyToProbe = createTestProbe[SequenceFactoryWorker.SequenceGenerated]()
      worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref)
      replyToProbe.expectNoMessage()

      // 再度予約が要求されてほしい
      inside(storeProbe.receiveMessage()) {
        case result: SequenceStore.ReserveSequence => // ここに来ない
      }

      testKit.stop(worker)
    }
```

Reply:
@tksugimoto `isEmpy` で `reserve` するよう修正しました。

5b3f53d1c

予約に失敗した場合は次の採番要求で予約をリトライする方針なので、仕様上 `reserve` する必要がありました。

`@SuppressWarnings(Array("lerna.warts.CyclomaticComplexity"))` がありますが、細かくメソッドに分割すると全体的な処理の流れを把握するのが難しくなる気がしたので、リファクタリングせずに残してあります。

Reply:
```scala
    if (nextSequence.isOverflow) {
      reset()
      empty(nextSequence)
    } else if (nextSequence.isEmpty) {
      val freeAmount = nextSequence.freeAmount
      if (freeAmount > 0) {
        reserve(sequenceContext = sequenceContext, amount = freeAmount)
        empty(nextSequence)
      } else {
        val message =
          s"freeAmount (${freeAmount.toString}) must be greater than 0 because freeAmount ≦ 0 means that the next sequence is overflow"
        logger.error(new IllegalStateException(message), message)
        Behaviors.stopped
      }
```

`maxSequenceValue === nextValue` の場合に、Actorが予期せず停止すると思われます

- isOverflow (`nextValue > config.maxSequenceValue`) ではない
- `freeAmount === ((config.maxSequenceValue - nextValue) / config.incrementStep).toInt === 0`

```scala
    "`maxSequenceValue === nextValue` の場合に、Actorが予期せず停止してしまう" in {
      val storeProbe        = createTestProbe[SequenceStore.Command]()
      val firstValue        = 3
      val incrementStep     = 10
      val reservationAmount = 1
      val worker = spawn(
        SequenceFactoryWorker
          .apply(
            maxSequenceValue = 13,
            firstValue = firstValue,
            incrementStep = incrementStep,
            reservationAmount = reservationAmount,
            storeProbe.ref,
            idleTimeout = 10.seconds,
            sequenceSubId,
          ),
      )

      val probe = akka.testkit.TestProbe()(system.classicSystem)
      probe.watch(worker.toClassic)

      // 初期化
      inside(storeProbe.receiveMessage()) {
        case result: SequenceStore.InitialReserveSequence =>
          result.replyTo ! SequenceStore.InitialSequenceReserved(
            initialValue = firstValue,                                                // 3
            maxReservedValue = firstValue + (incrementStep * (reservationAmount - 1)),// 3
          )
      }

      {
        val replyToProbe = createTestProbe[SequenceFactoryWorker.SequenceGenerated]()
        worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref)
        expect(replyToProbe.receiveMessage().value === BigInt(3)) // firstValue

        worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref)
      }

      probe.expectTerminated(worker.toClassic) // pass !!!
    }
```
※ 通ると良くないテストにしてあります

Reply:
@tksugimoto たしかに。
13 を予約しにいくのが期待する振る舞いのはずなので、`freeAmount` の計算に不備がありあそうです。
解決策検討してみます。

Reply:
> 13 を予約しにいくのが期待する振る舞いのはずなので、freeAmount の計算に不備がありあそうです。

13は初期化完了時に予約要求されています。

freeAmount  の計算はあっている気がします。

このPR前の処理だと、 `def reserving ` でしたが、なくなったので、 freeAmount ==0 がありえることになったのではないかと。

※ 2回目の採番要求は不要だったので訂正します。

```scala
      // 初期化
      inside(storeProbe.receiveMessage()) {
        case result: SequenceStore.InitialReserveSequence =>
          result.replyTo ! SequenceStore.InitialSequenceReserved(
            initialValue = firstValue,                                                // 3
            maxReservedValue = firstValue + (incrementStep * (reservationAmount - 1)),// 3
          )
      }

      println(storeProbe.receiveMessage())

      {
        val replyToProbe = createTestProbe[SequenceFactoryWorker.SequenceGenerated]()
        worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref)
        expect(replyToProbe.receiveMessage().value === BigInt(3)) // firstValue

//        worker ! SequenceFactoryWorker.GenerateSequence(sequenceSubId, replyToProbe.ref) // 不要だった
      }
```

Reply:
@tksugimoto 

> 13は初期化完了時に予約要求されています。

初期化完了時に予約済みなのは、3 ではないでしょうか？
`InitialSequenceReserved` の `maxReservedValue` は次のようになるはずです。
```
firstValue + (incrementStep * (reservationAmount - 1)
    3      +      10        * (  1               - 1) = 3
```

Reply:
@negokaz 
> 初期化完了時に予約済みなのは、3 ではないでしょうか？

それは、初期化時です。初期化完了時(初期化完了直後)は ↓の prepareNextSequence です。

```scala
    case WrappedSequenceStoreResponse(msg: SequenceStore.InitialSequenceReserved) =>
      logger.info("initial reserved: max:{}, initial:{}", msg.maxReservedValue, msg.initialValue)
      val sequenceContext: SequenceContext =
        SequenceContext(msg.maxReservedValue, nextValue = msg.initialValue)
      stashBuffer.unstashAll(prepareNextSequence(sequenceContext = sequenceContext, nextSequence = sequenceContext))
```
適当に print したログ。
```
(初期化レスポンスdone: ,3)
SequenceContext(3,3)
[isStarving] freeAmount: 1
(Reserving sequence: remain {}, add {}, current max reserved: {},0,1,3)
ReserveSequence(3,1,Some(test),Actor[akka://SequenceFactoryWorkerSpec/user/$a/$$a-adapter#1499095314]) // これが 13 予約
(SequenceGenerated: {},3)
SequenceContext(3,13)
isEmpty
freeAmount == 0
```

Reply:
`config.maxSequenceValue - nextValue` じゃなくて `config.maxSequenceValue - maxReservedValue` にすると、今回のケースに対応できてそうですね

Reply:
@tksugimoto `maxReservedValue` が 13 になってない状態で、`freeAmount == 0` になるのがまずいと思っています。
採番要求を受けたときに改めて 13 を予約しにいかないといけないはずなので。

次の修正により、上記のシナリオの最後の `freeAmount` が 0 ではなく、1 になります。

d5c6215a4 `🐞fix: maxSequenceValue を超えて予約しないように修正`